### PR TITLE
Updated gateway.py for MQTT Nodes

### PR DESCRIPTION
**Pull request by JSa1987**
New implementation of MQTT Nodes. All the code related to MQTT nodes is skipped if the MQTT nodes are not enabled. This replaces #225.

In addition, this fixes the issue discussed on #225 related to how out_id is handled when the initial relay synchronization is done.

@twa127 please review this in details and provide your feedback before this is merged into the main.

### CODE REVIEW

**Comment by JSa1987** [comment]
I forgot to mentioned that I tested this with my setup that includes:
- GPIO Relays
- MQTT Relays
- MySensors Sensors
- MQTT Sensors

**Comment by twa127** [comment]
I think this is a little bit better
[relay_init.txt](https://github.com/pihome-shc/PiHomeHVAC/files/7227975/relay_init.txt)

**Comment by twa127** [comment]
also, can we now drop the extra parameter added to set_relays() ?

**Comment by JSa1987** [comment]
@twa127 to make this work also for MQTT nodes we need to make the 4x adjustments I have highlighted below with **.
In short we cannot keep mixing out_id (ID of the record from the message_out table) with the Node ID.
```
    # re-sync any setup relays
    cur.execute("SELECT COUNT(*) FROM `relays`")
    count = cur.fetchone()
    count = count[0]
    # process any relays present
    if count > 0:
        cur.execute("SELECT distinct relays.`relay_id`, relays.`relay_child_id` , relays.`on_trigger` FROM `relays`, system_controller, zone_relays WHERE (relays.id = zone_relays.zone_relay_id) OR (relays.id = system_controller.heat_relay_id) OR (relays.id = system_controller.cool_relay_id) OR (relays.id = system_controller.fan_relay_id);")
        relays = cur.fetchall()
        relay_to_index = dict((d[0], i) for i, d in enumerate(cur.description))
        # get the last relay state from the messages_out table
        for x in relays:
            controler_id = x[relay_to_index["relay_id"]]
            out_child_id = x[relay_to_index["relay_child_id"]]
            out_on_trigger = x[relay_to_index["on_trigger"]]
            cur.execute(
                "SELECT `node_id`, `type` FROM `nodes` where id = (%s) LIMIT 1",
                (controler_id,),
            )
            nd = cur.fetchone()
            node_to_index = dict((d[0], i) for i, d in enumerate(cur.description))
            **node_out_id** = nd[node_to_index["node_id"]]
            node_type = nd[node_to_index["type"]]
            cur.execute(
                "SELECT COUNT(*) FROM `messages_out` where node_id = (%s) AND child_id = (%s) LIMIT 1",
                (**node_out_id**, out_child_id),
            )
            m_count = cur.fetchone()
            m_count = m_count[0]
            if m_count > 0:
                cur.execute(
                    "SELECT `id`, `sub_type`, `ack`, `type`, `payload` FROM `messages_out` where node_id = (%s) AND child_id = (%s) ORDER BY id DESC LIMIT 1",
                    (**node_out_id**, out_child_id),
                )
                msg = cur.fetchone()
                msg_to_index = dict((d[0], i) for i, d in enumerate(cur.description))
                out_id = msg[msg_to_index["id"]]
                out_sub_type = msg[msg_to_index["sub_type"]]
                out_ack = msg[msg_to_index["ack"]]
                out_type = msg[msg_to_index["type"]]
                out_payload = msg[msg_to_index["payload"]]
                msg =  str(out_id)  # Message ID
                msg += ";"  # Separator
                msg += str(out_child_id)  # Child ID of the Node.
                msg += ";"  # Separator
                msg += str(out_sub_type)
                msg += ";"  # Separator
                msg += str(out_ack)
                msg += ";"  # Separator
                msg += str(out_type)
                msg += ";"  # Separator
                msg += str(out_payload)  # Payload from DB
                msg += " \n"  # New line
                set_relays(
                    msg, node_type, out_id, out_child_id, out_on_trigger, out_payload, gatewayenableoutgoing, **out_node_id**
                )
        ping_timer = time.time()
    else:
        ping_timer = time.time()

```

If you agree with these adjustments I will incorporate this in the pull request.

Also, instead of performing two queries to the database wouldn't it be more efficient to use `cur.rowcount` ?
Maybe something like:

```
                cur.execute(
                    "SELECT `id`, `sub_type`, `ack`, `type`, `payload` FROM `messages_out` where node_id = (%s) AND child_id = (%s) ORDER BY id DESC LIMIT 1",
                    (**node_out_id**, out_child_id),
                )
                if cur.rowcount > 0:
                       msg = cur.fetchone()
                       ....
```

**Comment by JSa1987** [comment]
@twa127 For MQTT nodes we need the node_id in the set_relays() function to know which node to actuate. The parameters needed for MQTT nodes are:
- node_type       >> This is needed to flag the node as an MQTT node
- out_payload    >> This is needed to know if the ON or OFF command should be sent
- out_node_id    >> This together with out_child_id is needed to identify the relay that needs to be actuated
- out_child_id     >> This together with out_node_id is needed to identify the relay that needs to be actuated

**Comment by twa127** [comment]
the **  are purely cosmetic and yes agree with your second observation

**Comment by twa127** [comment]
if mqtt relays need the node_id, I'd much rather that it was obtained within the mqtt specific section of set_relays()

**Comment by JSa1987** [comment]
What would be the downside in passing the node_id as a parameter to the set_relays() function?
Have to retrieve this again inside the function with another query to the database is going to be inefficient.

The ** are to be able to pass the node_id to the set_relay() function, they are not cosmetic.

**Comment by twa127** [comment]
** what I meant is all you have done is change the wording, the original and your ** are exactly the same apart from the wording. 

as you can see from the MySensor, GPIO and Tasmota code sections of set_relays() they are self contained, you are passing in the node_id which is not required for the other 3 relay types, for systems with no mqtt relays this parameter is not required, so why generate and sent it when for 3 out of 4 relay types it is not required

**Comment by twa127** [comment]
I think the extra parameter thing is all a bit academic, I'm happy to go with the passing node_id, one suggestion please make it the 2nd and not the last parameter and call it node_id not out_node_id

**Comment by JSa1987** [comment]
@twa127 have a look now. I replaced out_node_id with node_id, made node_id the second parameter and included the IF to send the intitial synchronization message only if there is a record in the message_out table for the relay.

With regards to having parameters in the function used only for one type of node (i.e. node_id for MQTT nodes) this is a bit the same situation as having `msg` as parameter that is used only my the MySensor nodes. If we want to be very strict on this we probably should pass to the `set_relays()` function only `out_id` and obtain the other parameters needed in the function based on the type of node. But to be honest I'm happy with the current implementation.

**Comment by twa127** [comment]
I'm happy with the latest version, although I can't get 100% behind the extra parameter, you are 100% correct with your comment about about the msg parameter for MySensor devices. The only reason its there is that's how I started the development of set_relays(), in PiHome the other relay types were switched within boiler.php, the gateway handled the MySensor devices. With MaxAir all the switching is done by the gateway using the messages_out queue independent of relay type, I ported over the MySensor code and as it worked I left it as it was, in reality I should have built the message within set_relays()

Bottom line is I'm still not comfortable with the extra parameter, but I don't see the change impacting on the current functioning of the gateway, so lets get it deployed.

I do have some issues with the installation process but will raise them separately so as not delay progress anymore. 

Thanks for your efforts in adding a valuable addition to MaxAir.

**Comment by twa127** [comment]
see comments in issue #227

**Comment by JSa1987** [comment]
@twa127 with regards the extra parameter the way I see it is or we change everything and pass only the out_id and node_type parameters to the set_realys() function (and have the function retrieve what is needed based on node_type) or we leave it as it is now with msg, node_id and all the other parameters that are used only for some types of nodes.

With regards to the installation process to be honest I don't see this as an add-on but as a big enchantment of the basic functionality. Like other node types (MySensors, GPIO and Tasmota) that are supported by the MaxAir core and are not add-ons.
For MaxAir I believe that the future for wireless sensors and relays is Zigbee devices connected via [zigbee2mqtt](https://www.zigbee2mqtt.io/). Zigbee addresses the main issues we have seen with MySensors: unreliability of the connection and poor range (the zigbee network works as a mesh). In addition, this opens the door to using [hundreds](https://www.zigbee2mqtt.io/information/supported_devices.html) of cheap off the shelf devices. Example of interesting devices that will allow to define a zone for each room without major plumbing work are:
- [TuYa TS0601](https://www.zigbee2mqtt.io/devices/TS0601_thermostat.html) a smart TRV that allows to control radiators one by one in a boiler system. This one is the one I have been testing MQTT support with.
- [Keen Home SV01](https://www.zigbee2mqtt.io/devices/SV01.html) a smart HVAC vent that allows to open and close the vent itself.
My initial plan was to completely include this in the MaxAir core (without any installation) but the need of including an extra dependency cause some push back so I have put the dependency as add-on....

I completely agreed with all your comments in #227 regarding add-on such as Amazon Echo or Home Assistant Integration. These are not part of the core system and should be installed only if the end user wants to use them.

**Comment by JSa1987** [comment]
Updated 200921.sql to fix the issue reported by  @pihome-shc on #191

**Comment by twa127** [comment]
@JSa1987 I think I see it as both, the MQTT stuff is core and the HomeAssistant is an Add-On

**Comment by JSa1987** [comment]
@twa127 as mentioned in #212 for now this is just the backend for MQTT nodes support. The next step will be to adjust the web interface to be able to add the MQTT nodes and child from there. This is the next part I will be working on, but I expect to encounter some issue here as I have very limited php knowledge and no java script knowledge at all. So any help here would be more than well come. 
In short what would be needed is to update the Settings > Node and Zone Configuration > Nodes > Add node interface with an option for MQTT that when selected will show a couple more fields specific to MQTT nodes.

**Comment by twa127** [comment]
@JSa1987 we already have a way of manually adding nodes from the web interface just need to add an MQTT type, try adding <option value="MQTT">MQTT</option>
to modal.php at line 857
then Settings/Node and Zone Configuration/Nodes Add Node

**Comment by JSa1987** [comment]
@twa127 **that is exactly correct!**

While both the MQTT nodes and the Home Assistant Integration use the MQTT protocol, these are two very different things that should not be confused. The support for MQTT nodes should be a core functionality while the Home Assistant Integration should be a Add-on that the users that what to use it can install.

**Comment by twa127** [comment]
@JSa1987 what extra fields do you need?

**Comment by twa127** [comment]
@JSa1987 your change to 200921.sql will never get deployed no anyone who has already applied the original, that's not how the database update process works. You will need to create a new sql file which applies a change over the top of 200921.sql

**Comment by JSa1987** [comment]
@twa127 if I rename the new sql file as 250921.sql, would that be sufficient?

**Comment by JSa1987** [comment]
@pihome-shc I have added 250921.sql that fixes the you mentioned on #191. 
Can we now go ahead and merge this into Main?

**Comment by pihome-shc** [comment]
@JSa1987 can you please update [db_config.ini](https://github.com/pihome-shc/PiHomeHVAC/blob/main/st_inc/db_config.ini) 

build = 240921
db_build = 180921

**Comment by JSa1987** [comment]
I have updated db_config.ini to 250921
